File: data-pipeline-libraries-java/data-pipeline-gcp-observability-java/src/main/java/com/enrichmeai/culvert/gcp/observability/DataCatalogLineageEmitter.java
package com.enrichmeai.culvert.gcp.observability;

import com.enrichmeai.culvert.contracts.LineageEmitter;
import com.enrichmeai.culvert.lineage.LineageDestination;
import com.enrichmeai.culvert.lineage.LineageEvent;
import com.enrichmeai.culvert.lineage.LineagePipeline;
import com.enrichmeai.culvert.lineage.LineageSource;
import com.google.cloud.datacatalog.v1.CreateTagRequest;
import com.google.cloud.datacatalog.v1.DataCatalogClient;
import com.google.cloud.datacatalog.v1.Tag;
import com.google.cloud.datacatalog.v1.TagField;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * {@link LineageEmitter} implementation that writes lineage events as Data
 * Catalog tags on a configurable entry.
 *
 * <p>Each {@link LineageEvent} becomes one {@link Tag} attached to the
 * configured Data Catalog entry. The tag's fields mirror the four
 * sub-records of {@code LineageEvent} ({@code source}, {@code pipeline},
 * {@code destination}, {@code audit}), flattened to scalar string values.
 *
 * <h2>Why Data Catalog (and not Cloud Data Lineage)?</h2>
 *
 * <p>The newer Cloud Data Lineage API (artifact
 * {@code google-cloud-datalineage}) is not bundled in the GCP
 * {@code libraries-bom} this module pins. To avoid an out-of-BOM version
 * pin for a product still in transition, this Stage-2 implementation uses
 * the stable v1 {@link DataCatalogClient} and stores lineage as tags. A
 * dedicated {@code DataLineagePublisher} backed by the lineage API is
 * deferred to sprint-5 (tracked under the lineage epic).
 *
 * <h2>Construction</h2>
 *
 * <p>{@link #DataCatalogLineageEmitter(DataCatalogClient, String, String)}
 * — caller supplies an already-built client (so credentials, endpoint, and
 * lifecycle stay in the caller's hands), the fully-qualified Data Catalog
 * entry name to tag, and the tag template ID that defines the lineage tag
 * schema. The tag template must exist before {@link #emit(LineageEvent)} is
 * called; pre-provisioning is a Terraform / setup-time concern, not this
 * emitter's responsibility.
 *
 * <p>Implements {@link AutoCloseable}: {@link DataCatalogClient} is itself
 * closeable (extends {@link com.google.api.gax.core.BackgroundResource}),
 * so closing this emitter closes the wrapped client — matching the
 * {@code SecretManagerProvider} / {@code PubSubSource} pilot rule.
 *
 * <p>Sprint-2 deliverable for issue #24.
 */
public final class DataCatalogLineageEmitter implements LineageEmitter, AutoCloseable {

    private static final Logger LOG = LoggerFactory.getLogger(DataCatalogLineageEmitter.class);

    private final DataCatalogClient client;
    private final String entryName;
    private final String tagTemplate;

    /**
     * Primary constructor.
     *
     * @param client      Pre-built Data Catalog client. Required. Ownership
     *                    transfers to this emitter — {@link #close()} will
     *                    close it.
     * @param entryName   Fully-qualified Data Catalog entry name
     *                    ({@code projects/{p}/locations/{l}/entryGroups/{g}/entries/{e}})
     *                    to which lineage tags will be attached. Required.
     * @param tagTemplate Fully-qualified tag template name
     *                    ({@code projects/{p}/locations/{l}/tagTemplates/{t}})
     *                    that defines the lineage tag schema. Required.
     * @throws NullPointerException if any argument is null.
     */
    public DataCatalogLineageEmitter(DataCatalogClient client, String entryName, String tagTemplate) {
        this.client = Objects.requireNonNull(client, "client must not be null");
        this.entryName = Objects.requireNonNull(entryName, "entryName must not be null");
        this.tagTemplate = Objects.requireNonNull(tagTemplate, "tagTemplate must not be null");
    }

    /**
     * Emit one lineage event as a Data Catalog tag on the configured entry.
     *
     * <p>If the underlying Data Catalog call throws (network, auth, quota,
     * tag template mismatch), the exception is propagated to the caller.
     * Callers that batch their own events for at-least-once semantics
     * should catch and retry as appropriate.
     *
     * @param event The lineage event to emit. Required.
     * @throws NullPointerException if {@code event} is null.
     */
    @Override
    public void emit(LineageEvent event) {
        Objects.requireNonNull(event, "event must not be null");

        Tag.Builder tag = Tag.newBuilder()
                .setTemplate(tagTemplate);

        Map<String, String> fields = flatten(event);
        for (Map.Entry<String, String> e : fields.entrySet()) {
            tag.putFields(e.getKey(), TagField.newBuilder()
                    .setStringValue(e.getValue())
                    .build());
        }

        CreateTagRequest request = CreateTagRequest.newBuilder()
                .setParent(entryName)
                .setTag(tag.build())
                .build();

        client.createTag(request);
        LOG.debug("emitted lineage tag for entry {}", entryName);
    }

    /** The Data Catalog entry name lineage tags are attached to. */
    public String entryName() {
        return entryName;
    }

    /** The tag template the emitted tags reference. */
    public String tagTemplate() {
        return tagTemplate;
    }

    @Override
    public void close() {
        client.close();
    }

    /**
     * Flatten a {@link LineageEvent} into the scalar string-field shape
     * Data Catalog tags require. Empty optional sub-records are skipped so
     * the resulting tag stays compact.
     */
    private static Map<String, String> flatten(LineageEvent event) {
        Map<String, String> out = new HashMap<>();
        event.source().ifPresent(s -> putSource(out, s));
        event.pipeline().ifPresent(p -> putPipeline(out, p));
        event.destination().ifPresent(d -> putDestination(out, d));
        event.audit().ifPresent(a -> out.put("audit", a.toString()));
        return out;
    }

    private static void putSource(Map<String, String> out, LineageSource s) {
        out.put("source_type", s.type());
        out.put("source_uri", s.uri());
    }

    private static void putPipeline(Map<String, String> out, LineagePipeline p) {
        out.put("run_id", p.runId());
        out.put("pipeline_name", p.pipelineName());
        out.put("stage", p.stage());
        p.startedAt().ifPresent(t -> out.put("started_at", t.toString()));
        p.completedAt().ifPresent(t -> out.put("completed_at", t.toString()));
    }

    private static void putDestination(Map<String, String> out, LineageDestination d) {
        out.put("destination_type", d.type());
        out.put("destination_uri", d.uri());
    }
}

File: data-pipeline-libraries/data-pipeline-gcp-observability/src/data_pipeline_gcp_observability/data_catalog_lineage_emitter.py
"""DataCatalogLineageEmitter — LineageEmitter backed by Google Data Catalog.

Java sibling:
  data-pipeline-libraries-java/data-pipeline-gcp-observability-java/src/main/java/
  com/enrichmeai/culvert/gcp/observability/DataCatalogLineageEmitter.java

Imports ``google.cloud.datacatalog_v1`` lazily so the module can be imported
offline / in unit-test environments.  Pass a mock ``DataCatalogClient`` in
tests; never call the real API.

Sprint-19 / T19.2 — issue #125.
"""

from __future__ import annotations

import logging
from typing import Any

from data_pipeline_core.lineage.events import LineageEvent

logger = logging.getLogger(__name__)


class DataCatalogLineageEmitter:
    """LineageEmitter that writes lineage events as Data Catalog tags.

    Mirrors Java ``DataCatalogLineageEmitter`` (line 55).

    Each ``LineageEvent`` becomes one tag attached to the configured Data
    Catalog entry.  The tag's fields mirror the four sub-dicts of
    ``LineageEvent`` (source, pipeline, destination, audit), flattened to
    scalar string values — mirrors Java ``flatten()`` helper (line 137).

    Construction
    ------------
    - ``DataCatalogLineageEmitter(client, entry_name, tag_template)``
      Primary.  Mirrors Java constructor (line 77).  Caller owns the
      credential lifecycle; pass a ``MagicMock()`` for the client in tests.

    Lifecycle
    ---------
    Implements context-manager protocol and ``close()``.  Mirrors Java
    ``AutoCloseable.close()`` (line 128).
    """

    def __init__(self, client: Any, entry_name: str, tag_template: str) -> None:
        """Inject a pre-built DataCatalogClient and entry / template identifiers.

        Args:
            client:       DataCatalogClient (real or mock).  Required.
            entry_name:   Fully-qualified Data Catalog entry name
                          (``projects/{p}/locations/{l}/entryGroups/{g}/entries/{e}``).
                          Required.
            tag_template: Fully-qualified tag template name
                          (``projects/{p}/locations/{l}/tagTemplates/{t}``).
                          Required.

        Raises:
            TypeError: if any argument is None.
        """
        if client is None:
            raise TypeError("client must not be None")
        if entry_name is None:
            raise TypeError("entry_name must not be None")
        if tag_template is None:
            raise TypeError("tag_template must not be None")
        self._client = client
        self._entry_name = entry_name
        self._tag_template = tag_template

    # ------------------------------------------------------------------
    # LineageEmitter Protocol
    # ------------------------------------------------------------------

    def emit(self, event: LineageEvent) -> None:
        """Emit one lineage event as a Data Catalog tag — mirrors Java emit() (line 95).

        The event's sub-dicts are flattened to scalar string fields and written
        as a Data Catalog tag on ``entry_name`` using ``tag_template``.

        If the underlying Data Catalog call raises, the exception propagates to
        the caller (mirrors Java — exceptions are not swallowed here, unlike
        ``StageMetricsHook``).

        Args:
            event: The lineage event to emit.  Must not be None.

        Raises:
            TypeError: if ``event`` is None.
        """
        if event is None:
            raise TypeError("event must not be None")

        fields = self._flatten(event)

        # Lazy import — avoids hard dependency at module import time.
        try:
            from google.cloud import datacatalog_v1  # type: ignore[import]
        except ImportError as exc:  # pragma: no cover
            raise ImportError(
                "google-cloud-datacatalog is required for DataCatalogLineageEmitter.emit(). "
                "Install it with: pip install google-cloud-datacatalog"
            ) from exc

        tag = datacatalog_v1.Tag(template=self._tag_template)
        for field_name, str_value in fields.items():
            tag.fields[field_name] = datacatalog_v1.TagField(
                string_value=str_value
            )

        request = datacatalog_v1.CreateTagRequest(
            parent=self._entry_name,
            tag=tag,
        )
        self._client.create_tag(request=request)
        logger.debug("emitted lineage tag for entry %s", self._entry_name)

    # ------------------------------------------------------------------
    # Accessors / lifecycle
    # ------------------------------------------------------------------

    @property
    def entry_name(self) -> str:
        """Data Catalog entry name lineage tags are attached to — mirrors Java entryName() (line 117)."""
        return self._entry_name

    @property
    def tag_template(self) -> str:
        """Tag template the emitted tags reference — mirrors Java tagTemplate() (line 122)."""
        return self._tag_template

    def close(self) -> None:
        """Close the underlying DataCatalogClient — mirrors Java close() (line 128)."""
        try:
            self._client.close()
        except Exception:  # noqa: BLE001
            pass

    def __enter__(self) -> "DataCatalogLineageEmitter":
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _flatten(event: LineageEvent) -> dict[str, str]:
        """Flatten a LineageEvent into scalar string fields.

        Mirrors Java ``flatten()`` helper (line 137).  ``total=False``
        TypedDicts mean sub-dicts may be absent; absent keys are skipped.
        """
        out: dict[str, str] = {}

        source = event.get("source")
        if source:
            if "type" in source:
                out["source_type"] = str(source["type"])
            if "uri" in source:
                out["source_uri"] = str(source["uri"])

        pipeline = event.get("pipeline")
        if pipeline:
            if "run_id" in pipeline:
                out["run_id"] = str(pipeline["run_id"])
            if "pipeline_name" in pipeline:
                out["pipeline_name"] = str(pipeline["pipeline_name"])
            if "stage" in pipeline:
                out["stage"] = str(pipeline["stage"])
            if "started_at" in pipeline:
                out["started_at"] = str(pipeline["started_at"])
            if "completed_at" in pipeline:
                out["completed_at"] = str(pipeline["completed_at"])

        destination = event.get("destination")
        if destination:
            if "type" in destination:
                out["destination_type"] = str(destination["type"])
            if "uri" in destination:
                out["destination_uri"] = str(destination["uri"])

        audit = event.get("audit")
        if audit:
            out["audit"] = str(audit)

        return out
